src/core/experimental/array.d

```diff
@@ -0,0 +1,2299 @@
+        else static if (is(Payload[] : const(shared(void))[]))
+            alias Payload2Affix = shared Affix;
+        else static if (is(Payload[] : const(void)[]))
+            alias Payload2Affix = const Affix;
+        else
+            static assert(0, "Internal error for type " ~ Payload.stringof);
+    }
+
+    static auto ref prefix(T)(T[] b)
+    {
+        assert(b.ptr && (cast(size_t) b.ptr % alignment == 0));
+        return (cast(Payload2Affix!(T, size_t)*) b.ptr)[-1];
+    }
+
+    /**
+    Returns the global instance of this allocator type. The C heap allocator is
+    thread-safe, therefore all of its methods and `it` itself are `shared`.
+    */
+    static shared PrefixAllocator instance;
+}
+
+@safe pure nothrow @nogc unittest
+{
+    shared PrefixAllocator a;
+    auto b = a.allocate(42);
+    assert(b.length == 42);
+    assert(a.bytesUsed == 42);
+    () @trusted {
+        assert(a.prefix(b) == 0);
+        a.prefix(b)++;
+        assert(a.prefix(b) == 1);
+        a.deallocate(b);
+    }();
+    assert(a.bytesUsed == 0);
+}
+
+
+version(unittest)
+{
+    private alias SCAlloc = shared PrefixAllocator;
+    private alias SSCAlloc = shared PrefixAllocator;
+
+    SCAlloc localAllocator;
+    SSCAlloc sharedAllocator;
+
+    @nogc nothrow pure @trusted
+    void[] pureAllocate(bool isShared, size_t n)
+    {
+        return (cast(void[] function(bool, size_t) @nogc nothrow pure)(&_allocate))(isShared, n);
+    }
+
+    @nogc nothrow @safe
+    void[] _allocate(bool isShared, size_t n)
+    {
+        return isShared ? sharedAllocator.allocate(n) : localAllocator.allocate(n);
+    }
+
+    static if (__traits(hasMember, typeof(localAllocator), "expand"))
+    {
+        @nogc nothrow pure @trusted
+        bool pureExpand(bool isShared, ref void[] b, size_t delta)
+        {
+            return (cast(bool function(bool, ref void[], size_t) @nogc nothrow pure)(&_expand))(isShared, b, delta);
+        }
+
+        @nogc nothrow @safe
+        bool _expand(bool isShared, ref void[] b, size_t delta)
+        {
+            return isShared ?  sharedAllocator.expand(b, delta) : localAllocator.expand(b, delta);
+        }
+    }
+
+    @nogc nothrow pure
+    void pureDispose(T)(bool isShared, T[] b)
+    {
+        return (cast(void function(bool, T[]) @nogc nothrow pure)(&_dispose!(T)))(isShared, b);
+    }
+
+    @nogc nothrow
+    void _dispose(T)(bool isShared, T[] b)
+    {
+        return isShared ?  sharedAllocator.dispose(b) : localAllocator.dispose(b);
+    }
+}
+
+///
+struct Array(T)
+{
+private:
+
+    import core.atomic : atomicOp;
+
+    T[] payload;
+    Unqual!T[] support;
+
+    version(unittest)
+    {
+    }
+    else
+    {
+        alias localAllocator = shared PrefixAllocator.instance;
+        alias sharedAllocator = shared PrefixAllocator.instance;
+    }
+
+    static enum double capacityFactor = 3.0 / 2;
+    static enum initCapacity = 3;
+    bool isShared;
```

Comment:
eliminate

Reply:
Not sure if it can be eliminated as it's still needed in order to figure out if the prefix is a `shared size_t` or just a `size_t`.

Reply:
You can make it a tempalte parameter and require users to use helper functions to instatiate the type and make it's constructors private to the module.

Reply:
I thought about that, but I'd rather leave it as it is instead of having some wrappers. I was hoping I could get the compile time information about how the struct was instantiated.
I was hoping I can do something along the lines
```
static if (is(typeof(this) == immutable))
{
   // note that it started as immutable;
}
else
{
  // mutable reference
}
```

Such that
```
auto a = Array!int(1, 2, 3); // will hit the else branch
auto a2 = immutable Array!int(1, 2, 3); // will take the if branch
```

Reply:
I know, as I have found myself reaching for the same technique previously, but I also haven't found a reliable and clean way to do it. Let's hope that @andralex can offer an alternative.